### PR TITLE
Refactor logic for updating analyses

### PR DESCRIPTION
**Pull request by seallard**
## Description
Refactor logic for updating analyses with information from slurm. The logic does not belong in the data access layer.

New design
- Analysis service: uses a job service to update the analysis (status, progress etc.).
- Job service: uses a slurm service to retrieve information about jobs from slurm.
- Slurm service: defines an interface for retrieving information about jobs from slurm.
- Slurm CLI service: implements the slurm service interface, uses a Slurm CLI client.
- Slurm CLI client: retrieves information from slurm via the CLI (wraps the old logic).

I'll remove all the deprecated logic in a separate PR.

### Fixed
- Extract logic for parsing slurm from data access layer
- Abstract away slurm interactions to enable going via API or CLI

### This [version](https://semver.org/) is a
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

### CODE REVIEW

File: trailblazer/constants.py [hunk 1]
```diff
@@ -122,6 +126,10 @@ def statuses(cls) -> tuple:
     def ongoing_statuses(cls) -> tuple[str, str, str, str]:
         return cls.PENDING, cls.RUNNING, cls.COMPLETING, cls.ERROR
 
+    @classmethod
+    def fail_statuses(cls) -> tuple[str, str, str, str]:
+        return cls.ERROR, cls.FAILED
+
```
Comment (Vince-janv) [inline trailblazer/constants.py:132]: is this not defined on line 77? Also type hint should be `tuple[str, str]`

File: trailblazer/services/job_service/utils.py [hunk 2]
```diff
@@ -1,13 +1,45 @@
...
+def get_status(jobs: list[Job]) -> TrailblazerStatus:
+    is_running: bool = has_running_jobs(jobs)
+    is_failed: bool = has_failed_jobs(jobs)
+    if is_failed and is_running:
+        return TrailblazerStatus.ERROR
+    if is_failed:
+        return TrailblazerStatus.FAILED
+    if is_running:
+        return TrailblazerStatus.RUNNING
+    return TrailblazerStatus.CANCELLED
```
Comment (Vince-janv) [inline trailblazer/services/job_service/utils.py:27]: Is the old function not removed in this PR? I think there might be one piece of logic missing here:
```
    if single_analysis_status := _get_analysis_single_status(jobs_status_distribution):
        return single_analysis_status
```

File: trailblazer/services/slurm/slurm_service.py [hunk 3]
```diff
@@ -5,5 +5,9 @@
 
 class SlurmService(ABC):
     @abstractmethod
-    def get_job_info(self, job_id: int) -> SlurmJobInfo:
+    def get_job(self, job_id: int) -> SlurmJobInfo:
+        pass
+
+    @abstractmethod
+    def get_jobs(self, job_ids: list[int]) -> list[SlurmJobInfo]:
         pass
```
Comment (Vince-janv) [inline trailblazer/services/slurm/slurm_service.py:13]: Nice! Looking forward to seeing the REST-API service 😁

File: trailblazer/store/crud/update.py [hunk 4]
```diff
@@ -315,3 +317,16 @@ def update_user_token(self, refresh_token: str, user_id: int) -> None:
         user.refresh_token = refresh_token
         session: Session = get_session()
         session.commit()
+
+    def update_analysis_status(self, analysis_id: int, status: TrailblazerStatus) -> None:
+        analysis: Analysis | None = self.get_analysis_with_id(analysis_id)
+        analysis.status = status
+        session: Session = get_session()
+        session.commit()
```
Comment (Vince-janv) [inline trailblazer/store/crud/update.py:325]: This looks a bit strange, but I guess it's how you do it if you don't pass it as a parameter 🤔

File: trailblazer/services/job_service/utils.py (hunk 2 above)
Comment (seallard) [reply trailblazer/services/job_service/utils.py]: No, I'm leaving removing the deprecated code for a separate PR. It requires untangling a lot of other code.

Added the old logic here https://github.com/Clinical-Genomics/trailblazer/pull/423/commits/cf838778fcad580369e55230caf429304031bdf8.

File: trailblazer/store/crud/update.py (hunk 4 above)
Comment (seallard) [reply trailblazer/store/crud/update.py:325]: I agree, potentially we could just directly import the `Session` and use it as described here https://github.com/Clinical-Genomics/trailblazer/pull/423/commits/cf838778fcad580369e55230caf429304031bdf8.

I'll leave it for now, but might be something we want to revisit.

**Comment by seallard** [comment]
Tested in stage.

**Comment by seallard** [comment]
Deployed to production.

**Pull request merged**